Run GeneratorFilter modules in an external process controlled by cmsRun

Dr15Jones opened the pull request:
#### PR description:

ExternalGeneratorFilter is a cmsRun module which can be configured to run a module built from GeneratorFilter<> in a separate process. This separate process is controlled by ExternalGeneratorFilter and cooperatively shares the CPU with cmsRun.

#### PR validation:

Additional unit tests were added to compare the output of Pythia8GeneratorFilter run directly within cmsRun and run via the ExternalGeneratorFilter. With just 1 stream/thread the output is identical. With multiple streams, the GenLumiInfoHeader, edm::HepMCProduct, GenEventInfoProduct and the GenRunInfoProduct are identical. The GenLumiInfoProduct is nearly identical exception the XSec value and errors are different.

Dr15Jones committed:
Allow memory sharing between multiple ReadBuffers/WriteBuffers

The ReadBuffers can now safely determine if the underlying buffer
address was changed by a different ReadBuffer using the same
shared memory.

Dr15Jones committed:
Added ExternalGenerator*Info classes

These classes are used by the ExternalGeneratorFilter module for
communicating with an external process.

Dr15Jones committed:
First version of ExternalGeneratorFilter

Dr15Jones committed:
Fixed GenLumiInfoProduct::operator==

Identical objects were returned failed comparisons

Dr15Jones committed:
Added test to compare products from Generators

The added tests make sure what are supposed to be identical products
are actually identical.

Dr15Jones committed:
Treat 'missing' FinalStat special when adding

Dr15Jones committed:
Properly merge GenLumiInfoProduct at end LuminosityBlock

Also applied formatting.

Dr15Jones committed:
Allow XSec measurement to be different in comparisons

The code used to merge GenLumiInfoProducts does not handle merging
of the XSec information.

Dr15Jones committed:
Add verose option to external generator application

Can use a configuration parameter of ExternalGeneratorFilter to
turn on verbose reporting from the external process.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14624

- This PR adds an extra 92KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14624/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14624/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

Dr15Jones committed:
apply clang-tidy changes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14625

- This PR adds an extra 96KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

FWCore/Integration
FWCore/SharedMemory
GeneratorInterface/Core
GeneratorInterface/Pythia8Interface
SimDataFormats/GeneratorProducts


@SiewYan, @smuzaffar, @efeyazgan, @Dr15Jones, @mkirsano, @makortel, @cmsbuild, @agrohsje, @alberto-sanchez, @qliphy can you please review it and eventually sign? Thanks.
@mkirsano, @makortel, @rovere, @apsallid, @agrohsje, @wddgit, @lecriste, @alberto-sanchez, @fabiocos this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 9ff75a442ab4c5d12dc4d69b17127abfc5def516

CMSSW: CMSSW_11_1_X_2020-04-09-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5640/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

Dr15Jones committed:
Fix clang compiler warning

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14626

- This PR adds an extra 96KB to repository

cmsbuild commented:
Pull request #29445 was updated. @SiewYan, @smuzaffar, @efeyazgan, @Dr15Jones, @mkirsano, @makortel, @cmsbuild, @agrohsje, @alberto-sanchez, @qliphy can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 1015f46527e44bb7d6122bd2749c17ffc90a0ebf
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5641/summary.html
CMSSW: CMSSW_11_1_X_2020-04-09-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5641/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2694590
 - DQMHistoTests: Total failures: 51
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2694220
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

Dr15Jones commented:
For reviewers: the existing `GenLumiInfoProduct::mergeProduct` method does not attempt to combine the values for XSec. Instead, we just get the value from the first `GenLumiInfoProduct`. That seems less than optimal. Is there a reasonable way to combine the XSec values to get a more accurate number?

For reference, XSec generated for Pythia8 comes from here

https://github.com/cms-sw/cmssw/blob/cd9d8d7965adc4c0e9c885e3d64e9c48e9571382/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc#L626-L630

@smrenna any ideas?

Dr15Jones committed:
Properly handle dumpPython with ExternalGeneratorFilter

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14641

- This PR adds an extra 92KB to repository

cmsbuild commented:
Pull request #29445 was updated. @SiewYan, @smuzaffar, @efeyazgan, @Dr15Jones, @mkirsano, @makortel, @cmsbuild, @agrohsje, @alberto-sanchez, @qliphy can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
@dpiparo FYI

cmsbuild commented:
+1
Tested at: c56d07884971f0c051222cad898b92e4d3b2e4c7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5657/summary.html
CMSSW: CMSSW_11_1_X_2020-04-10-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5657/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2695166
 - DQMHistoTests: Total failures: 35
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2694812
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

Dr15Jones commented:
I did some simple speed comparisons. I started from a simple Pythia8 configuration
```python
process.generator = cms.EDFilter("Pythia8GeneratorFilter",
    comEnergy = cms.double(7000.),
    PythiaParameters = cms.PSet(
        pythia8_example02 = cms.vstring('HardQCD:all = on',
                                        'PhaseSpace:pTHatMin = 20.'),
        parameterSets = cms.vstring('pythia8_example02')
    )
```
This runs very fast (100,000 events in about 30 minutes) and has no filtering.

- If I use 1 thread and run the generator directly it takes 1815 seconds.
- If I use 1 thread and use the ExternalGeneratorFilter it takes 3035 seconds.
- If I use 4 threads and the ExternalGeneratorFilter it takes 903 seconds

The upshot is for this case where Pythia8 runs very fast per event, the time it takes to serialize/deserialize to/from the external application is a large overhead. However, it does scale better with more threads.

qliphy commented:
adding @jordan-martins  this should be good for BPH low filter efficiency requests.

qliphy commented:
> For reviewers: the existing `GenLumiInfoProduct::mergeProduct` method does not attempt to combine the values for XSec. Instead, we just get the value from the first `GenLumiInfoProduct`. That seems less than optimal. Is there a reasonable way to combine the XSec values to get a more accurate number?
> 
> For reference, XSec generated for Pythia8 comes from here
> 
> https://github.com/cms-sw/cmssw/blob/cd9d8d7965adc4c0e9c885e3d64e9c48e9571382/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc#L626-L630
> 
> @smrenna any ideas?

You may refer to GenXSecAnalyzer which combines GenLumiInfoProduct to get accurate Xsec.
https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Core/plugins/GenXSecAnalyzer.cc

Dr15Jones committed:
Also merge XSec during call to mergeProduct

This uses the calculation from GenXSecAnalyzer to do the weighted
averaging of XSec.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14658

- This PR adds an extra 96KB to repository

cmsbuild commented:
Pull request #29445 was updated. @SiewYan, @smuzaffar, @efeyazgan, @Dr15Jones, @mkirsano, @makortel, @cmsbuild, @agrohsje, @alberto-sanchez, @qliphy can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
>You may refer to GenXSecAnalyzer which combines GenLumiInfoProduct to get accurate Xsec.

I used that calculation to handle XSec improvements in call to GenLumiInfoProduct::mergeProduct.

cmsbuild commented:
+1
Tested at: 73a78a722dbd8aac6e62dd022068aeb981f5ac4b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5675/summary.html
CMSSW: CMSSW_11_1_X_2020-04-13-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5675/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2695166
 - DQMHistoTests: Total failures: 30
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2694817
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

makortel commented on GeneratorInterface/Core/bin/externalGenerator.cc:
```diff
@@ -0,0 +1,277 @@
+    std::string const uniqueID(vm[kUniqueIDOpt].as<std::string>());
+    {
+      //using namespace boost::interprocess;
+      //auto controlNameUnique = unique_name(memoryName, uniqueID);
```
Is this commented code still useful?

Dr15Jones replied:
I don't think so

Dr15Jones commented on GeneratorInterface/Core/bin/externalGenerator.cc:
```diff
@@ -0,0 +1,277 @@
+    std::string const memoryName(vm[kMemoryNameOpt].as<std::string>());
+    std::string const uniqueID(vm[kUniqueIDOpt].as<std::string>());
+    {
+      //using namespace boost::interprocess;
+      //auto controlNameUnique = unique_name(memoryName, uniqueID);
```
```suggestion
```

Dr15Jones committed:
Removed unneeded commented code

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14735

- This PR adds an extra 104KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  SimDataFormats/GeneratorProducts/src/classes.h  modified in PR(s): #29490
  - File  SimDataFormats/GeneratorProducts/src/classes_def.xml  modified in PR(s): #29490


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14735/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14735/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29445/14736

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  SimDataFormats/GeneratorProducts/src/classes.h  modified in PR(s): #29490
  - File  SimDataFormats/GeneratorProducts/src/classes_def.xml  modified in PR(s): #29490

cmsbuild commented:
Pull request #29445 was updated. @SiewYan, @smuzaffar, @efeyazgan, @Dr15Jones, @mkirsano, @makortel, @cmsbuild, @agrohsje, @alberto-sanchez, @qliphy can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ee434e678b3c40822c7c69cd293ec397042efda1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5739/summary.html
CMSSW: CMSSW_11_1_X_2020-04-16-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-60e73e/5739/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2696435
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2696114
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

makortel commented:
+core

makortel commented:
How can we eventually deploy the external GeneratorFilter? I'm wondering like what steps are needed from the GEN side, and where exactly we can or have to specify that a generator would be run externally (like does it have to into the generator configuration fragments, or can we have some more general switch e.g. in cmsDriver).

(I know this PR is probably not the best place to discuss if the questions are nontrivial, should we open an issue, or how should we proceed?)

Dr15Jones commented:
What is the plan for evaluating this? This was placed as a high priority by the L1s.

silviodonato commented:
> What is the plan for evaluating this? This was placed as a high priority by the L1s.

@alberto-sanchez @agrohsje @efeyazgan @mkirsano @qliphy @SiewYan

qliphy commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented on FWCore/SharedMemory/interface/ReadBuffer.h:
```diff
@@ -26,18 +26,19 @@ This works in conjunction with WriteBuffer.
   public:
     /** iUniqueName : must be unique for all processes running on a system.
-        iBufferIndex : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
+        iBufferInfo : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
```
Have you kept the same description on purpose?

Dr15Jones replied:
Yes

Dr15Jones replied:
The purpose of the variable is unchanged, I just changed the name to better reflect what it was doing.

silviodonato commented on FWCore/SharedMemory/interface/WriteBuffer.h:
```diff
@@ -29,21 +29,22 @@
   public:
     /** iUniqueName : must be unique for all processes running on a system.
-        iBufferIndex : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
+        iBufferInfo : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
```
same here

Dr15Jones replied:
That was intentional.

silviodonato commented:
+1
@Dr15Jones please have a look at my (trivial) comments above

agrohsje commented:
Hi @Dr15Jones @smrenna @qliphy ,
I thought about this a little bit. Wouldn't it make sense to combine the cross sections taking their relative uncertainties into account. Referring to this comment from Chris: 

> For reviewers: the existing `GenLumiInfoProduct::mergeProduct` method does not attempt to combine the values for XSec. Instead, we just get the value from the first `GenLumiInfoProduct`. That seems less than optimal. Is there a reasonable way to combine the XSec values to get a more accurate number?

makortel commented:
> I thought about this a little bit. Wouldn't it make sense to combine the cross sections taking their relative uncertainties into account. 

Hi @agrohsje, I believe @Dr15Jones already did that (see below)? Or is the logic in `GenXSecAnalyzer` sub-optimal?

https://github.com/cms-sw/cmssw/pull/29445#issuecomment-612929398
> > You may refer to GenXSecAnalyzer which combines GenLumiInfoProduct to get accurate Xsec.
> 
> I used that calculation to handle XSec improvements in call to GenLumiInfoProduct::mergeProduct.

srimanob commented:
Should we plan to do validation on high statistics relvals for the next pre-release? Could GEN come with few workflows to use this feature before?

agrohsje commented:
@makortel no. that is exactly what i had in mind. missed the commit from chris, sorry.

jordan-martins commented:
I did some preliminary test for the BPH dumby test as Chris used and further pointed me and it looks really nice. Previously, this dumby one had, with whatever number of cores (evtGen does not support MP), something around 35 events/lumi. With these implementations we could reach, with 4 cores, around 196 events/lumi. In terms of time/evt, prevously we had something around 0.3s to 0.6s. Now, my test indicates something around 0.0734s (a huge gain in core efficiency).

qliphy commented:
> Should we plan to do validation on high statistics relvals for the next pre-release? Could GEN come with few workflows to use this feature before?

We have done some local tests as you can see from above what Jordan reported on a BPH request, and also a MadGraph request:
https://github.com/cms-sw/cmssw/pull/28899#issuecomment-585828744

GEN plan to have a dedicated discussions on May 4th:
https://indico.cern.ch/event/894534/
where we will make further plans on validation, and foresee further improvement with MG support with readonly gridpack option.
